053-configurable-event-store/proposal.md

```diff
@@ -0,0 +1,457 @@
+# Summary
+
+This RFC proposes a way to support pluggable external stores for build events
+and to provide a general mechanism for processing build events.
+
+
+# Motivation
+
+The first (and primary) motivation for this change is to allow operators to
+configure external data store for build events. They may want to do this for a
+number of reasons:
+
+1. Builds can generate an enormous amount of build events - storing everything
+   in the same database as all of the other core-Concourse information means
+   that, over-time, the `build_events` tables will tend to dominate (in many
+   cases). For instance, in our (fairly small) Concourse deployment at
+   `ci.concourse-ci.org`, build events comprise ~70% of the total ~13GB used.
+   While storage is cheap, I can imagine this could result in general sluggishness
+   in larger Concourse deployments as indexes get large. While much of this can be
+   alleviated by more aggressive build log reaping, that can introduce performance
+   problems of its own, since deleting lots of build events can be taxing on
+   Postgres. I'm sure it's fairly obvious at this point that I really don't know
+   much about running Postgres/Concourse at scale, so let me know if what I'm
+   saying here doesn't hold true.
+1. Postgres wasn't designed for handling huge amounts of text data. Again, I
+   don't have much evidence/expertise to back this up, but having many large
+   rows in a `build_events` table strikes me as sub-optimal (compared to a system
+   designed for handling large amounts of text).
```

Comment:
Correct, PostgreSQL's text handling is somewhat suboptimal. But it's not _terrible_. Concourse is using it mostly for long-term storage of logs -- it's not being used as a fulltext search engine. In terms of storage, PostgreSQL automatically compresses `text` fields _if_ they are >2k long. That won't work for most events rows unfortunately.

What would make the biggest difference here would be a columnar store for the events table, since compression along a column of build logs is likely to be _very_ compressible indeed. The relevant 'zedstore' technology is being worked on but is a ways off, so irrelevant for this RFC.